feat(service-worker): add `openWindow` and `lastFocusedOrOpen` to not…

codingnuclei opened the pull request:
…ificationClick

Add `openWindow` and `lastFocusedOrOpen` abilty to the notificationClick handler such that when either a notification or notification action is clicked the service-worker can act accordingly without the need for the app to be open

PR Close #26907





## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number:  #26907 #29368

Currently the `notificationClick` handler in the service worker only broadcasts the event if the app is open.

The following enriched functionality is not present: 

Opening a new window on notification click
Focusing an already open window on click
Focusing an already open window on click and navigating to a new route

## What is the new behavior?

The notification now accepts a new data field `onActionClick`

```
notification: NotificationOptions = {
    "notification": {
        "title": "Now Fancy",       
		"actions": [
			{"action": "foo", "title": "Open new tab"},
			{"action": "bar", "title": "Focus last and Nav"},
			{"action": "baz", "title": "Focus last NO Nav"}
		],
        "data": {
			"onActionClick": {
			    "default": {"operation": "openWindow", "url": "/foo"},
				"foo": {"operation": "openWindow", "url": "/foo"},
				"bar": {"operation": "lastFocusedOrOpen", "navigate": true, "url": "/bar"},
				"baz": {"operation": "lastFocusedOrOpen", "navigate": false, "url": "/baz"}
			}
		},
        "body": "New notification"
    }
} 
```
The action will trigger the corresponding `onActionClick` based on a name match.


This PR handles the following operations:

`openWindow`: this will open a new window at the given `url` or at `/` if not provided

`lastFocusedOrOpen`:

1. If navigation true, will either open a new client or focus the latest client at the given url, depending on if there is a client to be focused
2. If navigation false, will focus the latest client but cause no navigation change. If no client, will open a new window at the provided url or default to `/`

If no action is clicked, but rather the notification itself, the default behavior is to open a new window at `/`, however this can be overwritten with a `default` option.


***Advice needed***

1. Advice on maybe where in the docs needs an update would be great.
~~2. I was unsure how to add test cases.  I can see there was ones that tested the swPush receiving the notification, but was struggling to think how to test browser tab opening etc~~



***Other***

~~I have also removed the custom types as ServiceWorker types are now provided by '--lib dom'~~

~~This caused a change required in other handlers to check if a client exists first as `this.scope.clients.get(clientId)` can now return a client or undefined~~

**EDIT**
Descoping type change as not part of scope of PR

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

codingnuclei committed:
feat(service-worker): add `openWindow`, `focusLastFocusedOrOpen` and `navigateLastFocusedOrOpen`

Add `openWindow`, `focusLastFocusedOrOpen` and `navigateLastFocusedOrOpen` abilty to the notificationclick handler such that when either a notification or notification action is clicked the service-worker can act accordinly without the need for the app to be open

PR Close #26907

codingnuclei commented:
Looks like tests are failing as many of the tests had the old service-worker typing.

If the concept of this PR looks good then I will go in and fix all these type errors

IgorMinar reviewed (commented):
thanks for the PR, @codingnuclei!

At a high level this change looks reasonable to me. 

@gkalpak can you please take a look at this one? 

(nit: if would be great to remove all of the unnecessary formatting changes to reduce the noise in the PR)

codingnuclei commented:
> (nit: if would be great to remove all of the unnecessary formatting changes to reduce the noise in the PR)

That is most of the formatting taken care of.  

clang-format formatted somethings differently, specifically when it came to the arrays in the specs.

Overall it is much cleaner now.

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -723,30 +684,242 @@ describe('Driver', () => {
+          });
+          expect(scope.clients.openWindow).toHaveBeenCalledWith('http://localhost:4200/fooz');
+        });
+        it('fallback to openWindow at `/`', async () => {
```
```suggestion

        it('falls back to openWindow at `/`', async () => {
```

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -723,30 +684,242 @@ describe('Driver', () => {
+                'default': {'operation': 'openWindow', 'url': 'fooz'},
+              },
+            },
+          });
```
AFAICT, (per the spec), this will be called with the empty string as the second argument when no specific action is clicked. I think it is worth testing that (instead or or in addition to `undefined`).

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -723,30 +684,242 @@ describe('Driver', () => {
+          await scope.handleClick({
+            title: 'This is a test without action',
+            body: 'Test body without action',
+            'data': {
+              'onActionClick': {},
```
```suggestion
            data: {
              onActionClick: {},
```

Although, I think it would be better to omit the `data` property altogether (to ensure everything still works fine).

gkalpak reviewed (changes requested):
Great work, @codingnuclei! I am really excited about this feature :tada:
Also, thank you for adding all those tests :100:

I've left a few comments/suggestions regarding the implementation. (Mainly preserving the current default behavior and splitting `focus` and `navigate` into two distinct operation types.)

BTW, we also need to document the new capabilities 📖 :wink:

codingnuclei commented on packages/service-worker/worker/src/driver.ts:
```diff
@@ -348,6 +348,38 @@ export class Driver implements Debuggable, UpdateSource {
+        break;
+      case 'lastFocusedOrOpen':
+        const windowClients = await this.scope.clients.matchAll({
+          type: 'window',
```
This was to do with my understanding of typings.

As matchAll could return WindowClient or Client I used the `type` option to help work out what the return type was.

Could possible use a type cast here instead.

More than happy to change if there is a better way (my typings skills arn't great 👍 )

gkalpak replied:
I see. You could get the typings work correctly by using an overloaded signature:

```
interface Clients {
  // ...
  matchAll<T extends Omit<ClientMatchOptions, 'type'>>(
    options?: T
  ): Promise<ReadonlyArray<WindowClient>>;
  matchAll<T extends ClientMatchOptions>(
    options: T
  ): Promise<ReadonlyArray<T['type'] extends 'window' | undefined ? WindowClient : Client>>;
  // ...
}
```

But I see that TypeScript's own [lib.webworker.d.ts][1] defines it the way you have it. So, let's stick to that (since it will make it easier to switch to the TS typings and get rid of ours). 

[1]: https://github.com/microsoft/TypeScript/blob/5be0d7156db35244ce2f3304550910251d51b9f6/lib/lib.webworker.d.ts#L950

codingnuclei commented on packages/service-worker/worker/src/driver.ts:
```diff
@@ -348,6 +348,38 @@ export class Driver implements Debuggable, UpdateSource {
+      case 'lastFocusedOrOpen':
+        const windowClients = await this.scope.clients.matchAll({
+          type: 'window',
+          includeUncontrolled: true,
```
This was based on my interpretation of the spec and also what is described here:
https://developers.google.com/web/fundamentals/push-notifications/common-notification-patterns

I guess in the case of angular/pwa; all clients will be controlled by the service worker, with no client using the same origin able to exist without the service-worker? 

With the above in mind `  includeUncontrolled: true,` is not required. Do you agree?

gkalpak replied:
TBH, I am not sure why that article suggests using `includeUncontrolled` (it doesn't give an explanation). Maybe it is suitable for the specific implementation used in the example.

It is not 100% clear by reading [the spec][1], but it sounds to me like `includeUncontrolled` would return clients that are on the same origin but outside the scope of the current SW, which is not what we want. Given that the Angular SW does claim all clients as soon as it is activated, I think it is safer to omit `includeUncontrolled` here, unless we manually test it and verify that it behaves correctly when registering multiple SWs on the same origin (with different scopes).

[1]: https://w3c.github.io/ServiceWorker/#clients-matchall

codingnuclei replied:
I have removed matching uncontrolled clients. As discussed I don't think it fits in the world of a SPA and angular

codingnuclei commented on packages/service-worker/worker/src/driver.ts:
```diff
@@ -348,6 +348,38 @@ export class Driver implements Debuggable, UpdateSource {
+        // As per the spec windowClients are `sorted in the most recently focused order`
+        let matchingClient: WindowClient|null = windowClients[0];
+        if (matchingClient) {
+          if (onActionClick.navigate) {
```
👍

codingnuclei commented on packages/service-worker/worker/src/driver.ts:
```diff
@@ -348,6 +348,38 @@ export class Driver implements Debuggable, UpdateSource {
+        }
+        break;
+      default:
+        await this.scope.clients.openWindow(urlToOpen);
```
👍

codingnuclei commented:
> BTW, we also need to document the new capabilities 📖 😉

Any ideas where this would best fit?

Somewhere in here: https://angular.io/api/service-worker/SwPush#notificationClicks  ?

codingnuclei commented on packages/service-worker/worker/src/driver.ts:
```diff
@@ -348,6 +348,38 @@ export class Driver implements Debuggable, UpdateSource {
+
+    const onActionClick = notification?.data?.onActionClick[notificationAction];
+
+    const urlToOpen = new URL(onActionClick?.url ?? '/', this.scope.location.origin).href;
```
if I understand you correctly then something like:

`    const urlToOpen = new URL('${onActionClick?.url}` ?? `${this.scope.location.pathname}/',  '${this.scope.location.origin}${this.scope.location.pathname}/}).href;`

gkalpak replied:
Not quite. `this.scope.location` has the path to the SW script which can be different than the scope of the SW. I meant something like this:

```suggestion
    const urlToOpen = new URL(onActionClick?.url ?? '', this.scope.registration.scope).href;
```

codingnuclei commented:
I have added some documentation for this PR

gkalpak commented on aio/content/guide/service-worker-notifications.md:
```diff
@@ -0,0 +1,121 @@
+
+### Actions
+
+Actions are a good way to expand how the end user interacts with the notification.
+
+Using the `onActionClick` field, each action can be tied to an `operation`:
```
Let's also mention `actions`. How about:

```suggestion
Actions offer a way to customize how the user can interact with a notification.

Using the `actions` property, you can define a set of available actions. Each action is represented as an action button that the user can click to interact with the notification.

In addition, using the `onActionClick` property on the `data` object, you can tie each action to an operation to be performed when the corresponding action button is clicked:
```

gkalpak commented on aio/content/guide/service-worker-notifications.md:
```diff
@@ -0,0 +1,121 @@
+      {"action": "foo", "title": "Open new tab"},
+      {"action": "bar", "title": "Focus last"},
+      {"action": "baz", "title": "Navigate last"},
+      {"action": "bazza", "title": "default openWindow"}
```
AFAICT, nothing will happen when 
```suggestion
      {"action": "qux", "title": "Just notify existing clients"}
```

gkalpak commented on aio/content/guide/service-worker-notifications.md:
```diff
@@ -0,0 +1,121 @@
+
+<div class="alert is-important">
+
+  If an action does not have a corresponding `onActionClick` field, then it will use the configured `default` `onActionClick`.
```
This is not true any more :grin:
Maybe instead mention what happens when the user clicks on the notification itself without clicking on a specific action.

gkalpak commented on aio/content/guide/service-worker-notifications.md:
````diff
@@ -0,0 +1,121 @@
+}
+```
+
+### Operations
````
```suggestion
#### Operations
```

gkalpak commented on aio/content/guide/service-worker-notifications.md:
```diff
@@ -0,0 +1,121 @@
+
+- `navigateLastFocusedOrOpen`: will focus the last focussed client and navigate that client to the given url relative to the service worker scope. If there is no client open, then will open a new client at the given url relative to the service worker scope
+
+### Actions
```
```suggestion
#### Actions
```

gkalpak commented on aio/content/guide/service-worker-notifications.md:
```diff
@@ -0,0 +1,121 @@
+
+<div class="alert is-important">
+
+  If a corresponding `onActionClick` field does not have a defined `url` then the service worker's registration scope will be used with no path.
```
I think this information should be moved to the `Operations` section.

gkalpak reviewed (changes requested):
Awesome! Thx for making the requested changes and adding the docs, @codingnuclei :tada:

Could you please also squash all commits into one (since it is one logical change/feature) and make any new changes on [fixup commits][1]? :pray:
(That makes reviewing new changes much easier.)

[1]: https://github.com/angular/angular/blob/a66dd8834c4ce27d317c5fc6e11fa38807747e9e/docs/FIXUP_COMMITS.md

codingnuclei committed:
fixup! feat(service-worker): add `openWindow`, `focusLastFocusedOrOpen` and `navigateLastFocusedOrOpen`

codingnuclei commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -723,30 +684,286 @@ describe('Driver', () => {
+    });
+
+    describe('Client interactions', () => {
+      describe('`openWindow` operation', () => {
```
I have changed all tests to use the `scope.registration.scope` when making assertions. I think this is a good change as now if the `SwTestHarnessBuilder.scopeUrl` changes, the tests will still pass.

codingnuclei commented:
> Awesome! Thx for making the requested changes and adding the docs, @codingnuclei 🎉
> 
> Could you please also squash all commits into one (since it is one logical change/feature) and make any new changes on [fixup commits](https://github.com/angular/angular/blob/a66dd8834c4ce27d317c5fc6e11fa38807747e9e/docs/FIXUP_COMMITS.md)? 🙏
> (That makes reviewing new changes much easier.)

All those changes have been made. Thanks for the feedback! 👍 🥇  

Sorry all that git noise must have made reviewing annoying. I have started using fixup commits.

I have resolved those review convos, but half way through doing it, it dawned on me that maybe I shouldn't have.

For future reference, what is recommend. The contributor or review to close the convos?

Sorry it made anything more of a hassle than it needed to be.

codingnuclei committed:
fixup! feat(service-worker): add `openWindow`, `focusLastFocusedOrOpen` and `navigateLastFocusedOrOpen`

gkalpak commented:
> Sorry all that git noise must have made reviewing annoying. I have started using fixup commits.

No worries. This has been a pleasure to review :wink:

> I have resolved those review convos, but half way through doing it, it dawned on me that maybe I shouldn't have.
> For future reference, what is recommend. The contributor or review to close the convos?

I've found it is subjective. Personally, I prefer for the reviewer to resolve them, but many people do it the other way around (the author resolves them). Since it is not possible to enforce one way or the other, it doesn't matter much, so feel free to do what works best for you :smiley:

mary-poppins commented:
You can preview f51a87f at https://pr42520-f51a87f.ngbuilds.io/.
You can preview 992bde3 at https://pr42520-992bde3.ngbuilds.io/.
You can preview 1284ac6 at https://pr42520-1284ac6.ngbuilds.io/.
You can preview d3efc80 at https://pr42520-d3efc80.ngbuilds.io/.
You can preview a06a38b at https://pr42520-a06a38b.ngbuilds.io/.
You can preview 78bcee2 at https://pr42520-78bcee2.ngbuilds.io/.
You can preview 1b5f2d9 at https://pr42520-1b5f2d9.ngbuilds.io/.
You can preview 0a13261 at https://pr42520-0a13261.ngbuilds.io/.
You can preview 0c06270 at https://pr42520-0c06270.ngbuilds.io/.
You can preview 1fb2636 at https://pr42520-1fb2636.ngbuilds.io/.
You can preview d00d3a7 at https://pr42520-d00d3a7.ngbuilds.io/.
You can preview 87ad5e1 at https://pr42520-87ad5e1.ngbuilds.io/.
You can preview d873a0e at https://pr42520-d873a0e.ngbuilds.io/.
You can preview 675744f at https://pr42520-675744f.ngbuilds.io/.

gkalpak reviewed (changes requested):
Thx again for making all the changes, @codingnuclei 🚀 
Another round of (mostly minor) comments. We're getting close :wink:

codingnuclei committed:
fixup! feat(service-worker): add `openWindow`, `focusLastFocusedOrOpen` and `navigateLastFocusedOrOpen`

codingnuclei committed:
fixup! feat(service-worker): add `openWindow`, `focusLastFocusedOrOpen` and `navigateLastFocusedOrOpen`

codingnuclei commented:
> Thx again for making all the changes, @codingnuclei 🚀
> Another round of (mostly minor) comments. We're getting close 😉

Got those in.  The checks are failing due to a rebase needing to be done. Will hold off till you give the go ahead 👍

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -723,30 +723,373 @@ describe('Driver', () => {
+
+      describe('URL resolution', () => {
+        it('should resolve relative to service worker scope', async () => {
+          scope = new SwTestHarnessBuilder().withServerState(server).build();
+          (scope.registration.scope as string) = 'http://localhost/foo/bar/';
+          driver = new Driver(scope, scope, new CacheDatabase(scope, scope));
```
Is it necessary to build new `scope` and `driver` instances? Couldn't you just update `scope.registration.scope` (or am I missing something)?

codingnuclei replied:
Absolutely right! :)

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -723,30 +723,373 @@ describe('Driver', () => {
+          expect(await makeRequest(scope, '/foo.txt')).toEqual('this is foo');
+
+          spyOn(scope.clients, 'openWindow');
+          const url = 'baz/qux';
```
There doesn't seem to be any benefit in declaring this variable here. Let's inline the value. (In the other two tests below as well.)

gkalpak commented on packages/service-worker/src/push.ts:
```diff
@@ -81,6 +81,7 @@ import {ERR_SW_NOT_SUPPORTED, NgswCommChannel, PushEvent} from './low_level';
  * <code-example path="service-worker/push/module.ts" region="subscribe-to-notification-clicks"
  * header="app.component.ts"></code-example>
  *
+ * @see {@link guide/service-worker-notifications Service worker notifications guide}
```
Maybe be more explicit here with something like:

```suggestion
You can read more on handling notification clicks in the [Service worker notifications guide](guide/service-worker-notifications).

```

gkalpak commented on aio/content/guide/service-worker-notifications.md:
```diff
@@ -0,0 +1,106 @@
+
+The default behaviour for the `notificationclick` event, is to close the notification and notify `SwPush.notificationClicks`.
+
+This can easily be changed by adding an `onActionClick` property to the `data` object, and provide a `default` operation.
```
Let's make it more clear that the `onActionClick` entries only configure _additional_ behavior (instead of changing the default behavior described above). How about:

```suggestion
You can specify an additional operation to be executed on `notificationclick` by adding an `onActionClick` property to the `data` object, and providing a `default` entry. This is especially useful for when there are no open clients when a notification is clicked.
```

gkalpak reviewed (changes requested):
Almost there :tada:
Yes, please rebase on latest master, @dario-piotrowicz :+1:

codingnuclei committed:
fixup! feat(service-worker): add `openWindow`, `focusLastFocusedOrOpen` and `navigateLastFocusedOrOpen`

codingnuclei committed:
fixup! feat(service-worker): add `openWindow`, `focusLastFocusedOrOpen` and `navigateLastFocusedOrOpen`

mary-poppins commented:
You can preview 41303ae at https://pr42520-41303ae.ngbuilds.io/.

gkalpak reviewed (approved):
Amazing job, @codingnuclei :tada:
Thank you for addressing all feedback so promptly and accurately :100:

LGTM :shipit:

(Leaving it to the docs team to give the final approval for the added docs.)

codingnuclei commented:
> Amazing job, @codingnuclei :tada:
> Thank you for addressing all feedback so promptly and accurately :100:
> 
> LGTM :shipit:
> 
> (Leaving it to the docs team to give the final approval for the added docs.)

Great! Cant wait to use it in prod!!

Has been my pleasure. Thank you for your time and effort getting this to the line.

TeriGlover commented on aio/content/guide/service-worker-notifications.md:
```diff
@@ -0,0 +1,106 @@
+
+<div class="alert is-important">
+
+  If an `onActionClick` item does not define a `url`, then the service worker's registration scope will be used.
```
```suggestion
  If an `onActionClick` item does not define a `url`, then the service worker's registration scope is used.
```
Removed passive voice.

TeriGlover commented on aio/content/guide/service-worker-notifications.md:
```diff
@@ -0,0 +1,106 @@
+
+#### Prerequisites
+
+A basic understanding of the following:
```
```suggestion
We recommend you have a basic understanding of the following:
```
Suggestion to form a complete sentence.

TeriGlover reviewed (approved):
Editing review complete; left a few comments, otherwise, LGTM.

codingnuclei committed:
fixup! feat(service-worker): add `openWindow`, `focusLastFocusedOrOpen` and `navigateLastFocusedOrOpen`

mary-poppins commented:
You can preview 8c0c3fc at https://pr42520-8c0c3fc.ngbuilds.io/.

codingnuclei commented:
That's the doc review changes in.  👍 

The failing checks are due to registry.yarnpkg.com "502 Bad Gateway". Can someone retrigger them please? :)

gkalpak reviewed (approved):
Reviewed-for: global-docs-approvers

AndrewKushnir reviewed (approved):
Reviewed-for: code-ownership

dylhunn commented:
Merged; thank you for your contribution!
